File: neetcode150/linked_lists/leetcode143.py
from typing import Optional

from data_structures.linked_list import ListNode
from utilities.linked_list import reverse


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head

        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        # now we have slow at half
        half = slow.next
        if slow.next:
            slow.next = None

        reversed_head = reverse(half)
        curr = head

        while curr:
            _next = curr.next
            curr.next = reversed_head
            reversed_head = _next

            curr = curr.next

        return head

File: js_algorithms/neetcode150/linked_lists/leetcode143.js
function ListNode(val, next) {
  this.val = val ?? 0;
  this.next = next ?? null;
}

/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
const reorderList = function (head) {
  // go through half of the list
  // reverse the second half of the list
  // once we do this, we would have two pointers
  // one from the beginning and another one from the end of the list
  let slow = head;
  let fast = head;

  // 1 -> 2 -> 3 -> 4
  // 1 -> 2 -> 3 -> 4 -> 5
  while (fast?.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  let otherHalf = null;

  if (slow?.next) {
    let temp = slow.next;
    slow.next = null;
    otherHalf = reverse(temp);
  }

  let result = head;

  while (result && otherHalf) {
    // store the next node
    let temp = result.next;
    // the next node should really be otherHalf
    result.next = otherHalf;
    otherHalf = temp;
    result = result.next;
  }
};

function reverse(head) {
  let prev = null;
  let curr = head;

  while (curr) {
    const next = curr.next;
    curr.next = prev;
    prev = curr;
    curr = next;
  }

  return prev;
}

module.exports = { reorderList };
